Mathlib/Tactic/StacksAttribute.lean

```diff
@@ -43,22 +43,59 @@ It extends the `Tag` environment extension with the data `declName, tag, comment
 def addTagEntry {m : Type → Type} [MonadEnv m] (declName : Name) (tag comment : String) : m Unit :=
   modifyEnv (tagExt.addEntry · { declName := declName, tag := tag, comment := comment })
 
-/--
-The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
-This allows `044Q3` and `GH3F6` as possibilities.
--/
-declare_syntax_cat stackTag
+open Parser
+
+/-- `stacksTag` is the node kind of Stacks Project Tags: a sequence of digits and
+uppercase letters. -/
+abbrev stacksTagKind : SyntaxNodeKind := `stacksTag
+
+/-- The main parser for Stacks Project Tags: it accepts any letter or digit, even though
+tags do not allow lowercase letters.
+This allows for better error messages raised by the elaborator. -/
+def stacksTagFn : ParserFn := fun c s =>
+  let i := s.pos
+  let st := takeWhileFn (·.isAlphanum) c s
+  mkNodeToken stacksTagKind i c st
+
+@[inherit_doc stacksTagFn]
+def stacksTagNoAntiquot : Parser := {
+  fn   := stacksTagFn
+  info := mkAtomicInfo "stacksTag"
+}
+
+@[inherit_doc stacksTagFn]
+def stacksTag : Parser :=
+  withAntiquot (mkAntiquot "stacksTag" stacksTagKind) stacksTagNoAntiquot
+
+end Mathlib.Stacks
+
+namespace Lean.PrettyPrinter
+open Mathlib.Stacks
 
-@[inherit_doc Parser.Category.stackTag]
-syntax (num)? (ident)? : stackTag
+namespace Formatter
+
+/-- The formatter for Stacks Project Tags syntax. -/
+@[combinator_formatter stacksTagNoAntiquot] def stacksTagNoAntiquot.formatter :=
+  visitAtom stacksTagKind
+
+end Formatter
+
+namespace Parenthesizer
+
+/-- The parenthesizer for Stacks Project Tags syntax. -/
+@[combinator_parenthesizer stacksTagNoAntiquot] def stacksTagAntiquot.parenthesizer := visitToken
+
+end Lean.PrettyPrinter.Parenthesizer
+
+namespace Mathlib.Stacks
 
 /-- The `stacks` attribute.
 Use it as `@[stacks TAG "Optional comment"]`.
-The `TAG` is mandatory.
+The `TAG` is mandatory and should be a sequence of 4 digits or uppercase letters.
 
 See the [Tags page](https://stacks.math.columbia.edu/tags) in the Stacks project for more details.
 -/
-syntax (name := stacks) "stacks " (stackTag)? (ppSpace str)? : attr
+syntax (name := stacks) "stacks " (stacksTag)? (ppSpace str)? : attr
```

Comment:
```suggestion
syntax (name := stacks) "stacks " stacksTag (ppSpace str)? : attr
```
I think we should require the stacks tag.

Reply:
Actually, I am not sure if this change does anything?

Reply:
I *think* that it does nothing, since the empty stacksTag is allowed.  This probably also explains the unreachable code below.

Reply:
You can get a slightly better parser error with 
```lean
syntax (name := stacks) "stacks " stacksTag !(ident) (ppSpace str)? : attr
```
which replaces "unexpected identifier" with "unexpected element" if you only use digits and uppercase letters in `stacksTag`. 

Parsers themselves have errors built and aesthetically validation *should* be built into the parser but I am not sure of the recipe to get custom error here.